### DISCUSSION TITLE
Add direct support for SSL/TLS

### DISCUSSION THREAD

**Discussion by Caligatio** [mealie-recipes/mealie#2294] [Feature Request]
**Before submitting this feature request I have**
- [x] Reviewed the documentation and verified that my feature request isn't already a feature.
- [x] Reviewed the existing feature requests and my feature has not been requested.
- [x] Checked the [tasks taged](https://github.com/hay-kot/mealie/issues?q=is%3Aissue+is%3Aopen+label%3Atask+) issues and verified my feature is not covered

**Please Describe The Problem To Be Solved**
Mealie currently lacks direct support for serving itself via SSL/TLS. Given that certain features do not work without SSL/TLS (e.g. copying public link to the clipboard), this seems like a hybrid bug/feature request.

I currently have a Caddy reverse proxy server that makes internal web services externally available. To make this work with Mealie, I have certs issued to the Caddy server but I also then need to throw nginx in front of Mealie with its own certs. The nginx server is only needed because Mealie doesn't support SSL. I recently found out there are a few gotchas using nginx this way ([variables vs hostnames](https://serverfault.com/questions/240476/how-to-force-nginx-to-resolve-dns-of-a-dynamic-hostname-everytime-when-doing-p) and [DNS TTL](https://github.com/moby/moby/issues/30029)) that caused me some issues.

**(Optional): Suggest A Solution**
It looks trivial to add SSL/TLS support to a [uvicorn instance](https://www.uvicorn.org/deployment/#running-with-https).

**Additional Information**
- [x] If the feature is accepted I'm willing to submit a PR to provide this feature

**Comment by cmintey** [comment]
I don't fully understand why you need the nginx server between Caddy and Mealie.

I also am running Mealie behind Caddy reverse proxy over SSL, and I can use those features such as copy to clipboard. Could it be a misconfiguration in your Caddy proxy?

**Comment by Caligatio** [comment]
I should have specified a bit better: LAN clients bypass Caddy and go directly to nginx/Mealie whereas external clients go to the Caddy server (in a firewall DMZ) which then proxies to nginx.

**Reply by cmintey** [reply]
Okay I see. I know Caddy does support [internal HTTPS](https://caddyserver.com/docs/automatic-https#local-https), so a potential solution is to have a reverse proxy listen on `mealie.localhost`. Your LAN clients would connect through Caddy, but wouldn't be leaving your network.

However, I can also see that serving SSL/TLS from the Mealie container could be beneficial to many users.

**Reply by LargoUsagi** [reply]
You can also split horizon your DNS records so that your LAN clients see a route to the proxy locally and then talk with that instead of the mealie deployment directly.

**Comment by LargoUsagi** [comment]
I would advise against putting TLS inside of the mealie containers, certificate rotation inside of containers is god awful.  External service termination using tools like nginx, caddy, cloud vendor API gateways and the like are all normal and current practices.

**Comment by Caligatio** [comment]
I would also point out that HTTP3 and many implementations of HTTP2 require SSL/TLS. I don't believe that gunicorn supports HTTP2 yet so it's not an immediate concern. Also, many new HTML features (clipboard, service workers, etc) require SSL/TLS to work.

I would agree that something like Caddy that manages certificates is the better solution but I also feel like it should be possible to run an application (with all its features) without a reverse proxy. To borrow a popular phrase, don't let perfect be the enemy of good.
